File: newlib/libc/sys/redox/folder.c
#include "common.h"

#include <sys/dirent.h>

DIR * _opendir(const char * path) {
    int fd = _open(path, O_RDONLY | O_CLOEXEC | O_DIRECTORY, 0);

    if(fd >= 0){
        DIR * dir = (DIR *)calloc(sizeof(DIR), 1);
        dir->dd_fd = fd;
        return dir;
    }

    return NULL;
}

struct dirent * _readdir(DIR * dir){
    if(dir){
        //TODO: Speed improvements
        int i;
        for(i = 0; i < 4096; ++i){
            if(_read(dir->dd_fd, &(dir->dd_ent.d_name[i]), 1) > 0){
                if(dir->dd_ent.d_name[i] == '\n'){
                    break;
                }
            }else{
                break;
            }
        }
        dir->dd_ent.d_name[i] = '\0';

        if(i > 0){
            return &(dir->dd_ent);
        }
    }

    return NULL;
}

void _rewinddir(DIR * dir){
    if(dir){
        _lseek(dir->dd_fd, 0, 0);
    }
}

int _closedir(DIR * dir){
    if(dir){
        _close(dir->dd_fd);
        free(dir);
        return 0;
    }

    return -1;
}

File: newlib/libc/sys/redox/src/folder.rs
use ::{c_int, c_char};
use syscall::{self, O_CLOEXEC, O_RDONLY, O_DIRECTORY};
use core::ptr::null;
use alloc::boxed::Box;
use ::file::PATH_MAX;

#[repr(C)]
pub struct dirent {
    pub d_name: [c_char; PATH_MAX],
}

pub struct DIR {
    pub fd: ::RawFile,
    pub ent: dirent,
    pub buf: [u8; PATH_MAX],
    pub count: usize,
    pub pos: usize
}

libc_fn!(unsafe opendir(path: *mut c_char) -> Result<*mut DIR> {
    let path = ::cstr_to_slice(path);
    let fd = ::RawFile::open(path, O_RDONLY | O_CLOEXEC | O_DIRECTORY)?;
    let dir = Box::new(DIR {
        fd,
        ent: dirent { d_name: [0; PATH_MAX] },
        buf: [0; PATH_MAX],
        count: 0,
        pos: 0

    });
    Ok(Box::into_raw(dir))
});

libc_fn!(unsafe readdir(dir: *mut DIR) -> Result<*const dirent> {
    if let Some(dir) = dir.as_mut() {
        let mut i = 0;
        'outer: while i < PATH_MAX - 1 {
            while dir.pos < dir.count {
                dir.ent.d_name[i] = dir.buf[dir.pos] as c_char;
                dir.pos += 1;
                if dir.buf[dir.pos-1] == b'\n' {
                    break 'outer;
                }
                i += 1;
            }
            dir.count = syscall::read(*dir.fd, &mut dir.buf)?;
            if dir.count == 0 {
                break;
            }
            dir.pos = 0;
        }
        if i != 0 {
            dir.ent.d_name[i] = 0;
            return Ok(&dir.ent);
        }
    }
    Ok(null())
});

libc_fn!(unsafe rewinddir(dir: *mut DIR) {
    if let Some(dir) = dir.as_mut() {
        dir.count = 0;
        let _ = syscall::lseek(*dir.fd, 0, syscall::SEEK_SET);
    }
});

libc_fn!(unsafe closedir(dir: *mut DIR) -> Result<c_int> {
    Box::from_raw(dir);
    Ok(0)
});
